File: bot_engine/notify/log.py
"""Log-only notification backend (no-op fallback)."""

from __future__ import annotations

import logging

from bot_engine.notify.base import NotifyBackend

log = logging.getLogger(__name__)


class LogNotifyBackend(NotifyBackend):
    """Logs messages instead of sending them. Used when no platform is configured."""

    def __init__(self, default_channel: str = "") -> None:
        self._default_channel = default_channel

    def send_message(
        self,
        channel: str,
        text: str,
        *,
        thread_ts: str = "",
        unfurl_links: bool = False,
    ) -> str:
        log.info(
            "[notify:log] channel=%s text=%s",
            channel or self._default_channel or "(none)",
            text[:200],
        )
        return ""

    def add_reaction(self, channel: str, timestamp: str, emoji: str) -> None:
        log.debug("[notify:log] reaction=%s ts=%s", emoji, timestamp)

File: go/internal/notify/log.go
package notify

import (
	"context"
	"log/slog"
)

// LogNotifier logs notifications instead of sending them.
// Used as a fallback when no platform credentials are configured.
type LogNotifier struct{}

// NewLogNotifier creates a LogNotifier.
func NewLogNotifier() *LogNotifier {
	return &LogNotifier{}
}

func (l *LogNotifier) SendMessage(_ context.Context, channel, text, _ string) (string, error) {
	slog.Info("notify/log: message", "channel", channel, "text", truncate(text, 200))
	return "", nil
}

func (l *LogNotifier) AddReaction(_ context.Context, channel, timestamp, emoji string) error {
	slog.Debug("notify/log: reaction", "emoji", emoji, "channel", channel, "ts", timestamp)
	return nil
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}
